frontend/src/features/admin-form/responses/IndividualResponsePage/RenderedSignatureCanvas.tsx

```diff
@@ -0,0 +1,118 @@
+   * If not provided, the signature will be scaled to fit the container.
+   */
+  widthPx?: number
+}
+
+export const RenderedSignatureCanvas = ({
+  row,
+  widthPx,
+}: DecryptedRowBaseProps & RenderedSignatureCanvasSizeProps): JSX.Element => {
+  const canvasRef = useRef<HTMLCanvasElement | null>(null)
+  const [containerRef, { width: containerWidth }] = useMeasure<HTMLDivElement>()
+  const [img, setImg] = useState<string | null>(null)
+  const [canvasSize, setCanvasSize] = useState({ width: 0, height: 0 })
+
+  useEffect(() => {
+    const canvas = canvasRef.current
+    if (!canvas) return
+
+    const vectorArray: SignatureVectorArray =
+      row.answerArray && row.answerArray[1]
+        ? convertToSignatureVectorArray(row.answerArray[1] as string)
+        : []
+
+    const ctx = canvas.getContext('2d')
+    if (!ctx) return
+
+    // Clear canvas first
+    ctx.clearRect(0, 0, canvas.width, canvas.height)
+
+    if (vectorArray.length === 0) {
+      // No signature - shrink canvas to minimal size
+      canvas.width = 0
+      canvas.height = 0
+      setCanvasSize({ width: 0, height: 0 })
+      return
+    }
+
+    const { minX, minY, maxX, maxY } = getBoundingBox(vectorArray)
+    const padding = SIGNATURE_OUTPUT_PADDING_DEFAULT
+
+    // apply devicePixelRatio to maintain sharpness
+    const dpr = window.devicePixelRatio || 1
+    canvas.width = maxX * dpr
+    canvas.height = maxY * dpr
+    // keep the CSS size same as logical size
+    canvas.style.width = `${maxX}px`
+    canvas.style.height = `${maxY}px`
+
+    // Scale the context to account for dpr
+    ctx.scale(dpr, dpr)
+    setCanvasSize({ width: maxX, height: maxY })
+
+    // Draw strokes using original coordinates but shift by minX and minY to remove whitespace
+    vectorArray.forEach((stroke) => {
+      const shiftedStroke = stroke.map(([x, y, pressure]) => [
+        x - minX + padding,
+        y - minY + padding,
+        pressure,
+      ])
+      const pathData = getStroke(shiftedStroke, {
+        size: SIGNATURE_STROKE_SIZE,
+        thinning: SIGNATURE_STROKE_THINNING,
+        smoothing: SIGNATURE_STROKE_SMOOTHING,
+        streamline: SIGNATURE_STROKE_STREAMLINE,
+      })
+      drawStroke(ctx, pathData)
+    })
+
+    // Convert to image so that it is responsive to screen size and
+    // maintains sharpness of original canvas
+    const img = canvas.toDataURL('image/png', 1)
+    setImg(img)
+  }, [row.answerArray])
+
+  return (
+    <Box
+      ref={containerRef}
+      background="white"
+      width={widthPx ?? '100%'}
+      maxW={canvasSize.width}
+      maxH={canvasSize.height}
+      borderColor="neutral.400"
+      borderRadius="0.25rem"
+    >
+      {img && (
+        <img
+          width={widthPx ?? containerWidth}
+          style={{ maxWidth: canvasSize.width, maxHeight: canvasSize.height }}
+          src={img}
+        />
+      )}
+      <canvas ref={canvasRef} style={{ display: 'none' }} />
+    </Box>
+  )
```

Comment:
I think we need to conditionally render this canvas or omit this canvas from the printed PDF.

```
Uncaught InvalidStateError: Failed to execute 'drawImage' on 'CanvasRenderingContext2D': The image argument is a canvas element with a width or height of 0.
```

<img width="1624" height="978" alt="image" src="https://github.com/user-attachments/assets/8e9fb49f-4d57-461a-ab97-7143616e9a42" />

Reply:
thanks @eliotlim for pointing this out - let me have a look and see how to resolve this

Reply:
fixed in [e60e7eb](https://github.com/opengovsg/FormSG/pull/8700/commits/e60e7eb62b0d1fc3f406f84d97a84dd043707f7a)!

Reply:
added tc to cover it in [b9daf80](https://github.com/opengovsg/FormSG/pull/8700/commits/b9daf8000d57411757e5f009fb7b0eb25cf211a4)